File: backend/app/api/tasks.py
"""Public Task API routes for CP4.

These endpoints expose TaskService without changing the legacy session/chat API.
They operate on app-owned task metadata only; Hermes execution remains behind the
existing backend policy boundary.
"""
from __future__ import annotations

import json
from typing import Annotated

import aiosqlite
from fastapi import APIRouter, Depends, Header, HTTPException, Query, Response, status
from sse_starlette.sse import EventSourceResponse

from app.api.schemas import (
    PublicTaskActionCreateRequest,
    PublicTaskActionDecisionRequest,
    PublicTaskActionResponse,
    PublicTaskCreateRequest,
    PublicTaskEventResponse,
    PublicTaskResponse,
)
from app.dependencies import get_db
from app.repositories.idempotency_repository import IdempotencyConflict
from app.repositories.task_repository import TaskRepository
from app.services.audit import log_audit_event
from app.services.state_machine import TransitionError
from app.services.task_service import TaskService

router = APIRouter(prefix="/api/tasks", tags=["Tasks"])


def _task_response(task: dict, duplicate: bool = False) -> PublicTaskResponse:
    data = dict(task)
    data["duplicate"] = duplicate
    return PublicTaskResponse(**data)


def _action_response(action: dict) -> PublicTaskActionResponse:
    return PublicTaskActionResponse(**dict(action))


async def _get_task_or_404(db: aiosqlite.Connection, task_id: str) -> dict:
    task = await TaskRepository(db).get_task(task_id)
    if task is None:
        raise HTTPException(status_code=404, detail="Task not found")
    return task


@router.post("", response_model=PublicTaskResponse, status_code=status.HTTP_201_CREATED)
async def create_task(
    request: PublicTaskCreateRequest,
    response: Response,
    conn: aiosqlite.Connection = Depends(get_db),
    idempotency_key: Annotated[str | None, Header(alias="Idempotency-Key")] = None,
) -> PublicTaskResponse:
    """Create a public task record.

    Idempotency-Key is optional. If provided, the same key with the same payload
    returns the original task; the same key with a different payload returns 409.
    """
    service = TaskService(conn)
    try:
        task, duplicate = await service.create_task(
            session_id=request.session_id,
            title=request.title,
            description=request.description,
            task_type=request.task_type,
            parent_task_id=request.parent_task_id,
            idempotency_key=idempotency_key,
        )
    except IdempotencyConflict as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc

    await log_audit_event(
        conn=conn,
        session_id=request.session_id,
        actor="api",
        action="task.created" if not duplicate else "task.create_replayed",
        target=task["id"],
        payload={
            "idempotency_key": bool(idempotency_key),
            "duplicate": duplicate,
            "task_type": request.task_type,
        },
    )
    await conn.commit()
    if duplicate:
        response.status_code = status.HTTP_200_OK
    return _task_response(task, duplicate=duplicate)


@router.get("", response_model=list[PublicTaskResponse])
async def list_tasks(
    conn: aiosqlite.Connection = Depends(get_db),
    session_id: str | None = None,
    status_filter: Annotated[str | None, Query(alias="status")] = None,
    limit: int = Query(50, ge=1, le=200),
    offset: int = Query(0, ge=0),
) -> list[PublicTaskResponse]:
    tasks = await TaskRepository(conn).list_tasks(
        session_id=session_id,
        status=status_filter,
        limit=limit,
        offset=offset,
    )
    await log_audit_event(
        conn=conn,
        session_id=session_id,
        actor="api",
        action="task.listed",
        payload={"status": status_filter, "limit": limit, "offset": offset},
    )
    await conn.commit()
    return [_task_response(task) for task in tasks]


@router.get("/{task_id}", response_model=PublicTaskResponse)
async def get_task(
    task_id: str,
    conn: aiosqlite.Connection = Depends(get_db),
) -> PublicTaskResponse:
    task = await _get_task_or_404(conn, task_id)
    await log_audit_event(
        conn=conn,
        session_id=task.get("session_id"),
        actor="api",
        action="task.viewed",
        target=task_id,
    )
    await conn.commit()
    return _task_response(task)


@router.get("/{task_id}/events", response_model=list[PublicTaskEventResponse])
async def list_task_events(
    task_id: str,
    conn: aiosqlite.Connection = Depends(get_db),
) -> list[PublicTaskEventResponse]:
    task = await _get_task_or_404(conn, task_id)
    events = await TaskRepository(conn).get_events(task_id)
    await log_audit_event(
        conn=conn,
        session_id=task.get("session_id"),
        actor="api",
        action="task.events_listed",
        target=task_id,
        payload={"count": len(events)},
    )
    await conn.commit()
    return [PublicTaskEventResponse(**event) for event in events]


@router.get("/{task_id}/events/stream")
async def stream_task_events(
    task_id: str,
    conn: aiosqlite.Connection = Depends(get_db),
) -> EventSourceResponse:
    task = await _get_task_or_404(conn, task_id)
    events = await TaskRepository(conn).get_events(task_id)
    await log_audit_event(
        conn=conn,
        session_id=task.get("session_id"),
        actor="api",
        action="task.events_streamed",
        target=task_id,
        payload={"count": len(events)},
    )
    await conn.commit()

    async def _generate():
        for event in events:
            yield {
                "event": event["type"],
                "data": json.dumps(event, ensure_ascii=False),
            }
        if task["status"] in {"succeeded", "failed", "cancelled"}:
            yield {
                "event": "done",
                "data": json.dumps({"type": "done", "task_id": task_id}),
            }

    return EventSourceResponse(_generate())


@router.post("/{task_id}/start", response_model=PublicTaskResponse)
async def start_task(
    task_id: str,
    conn: aiosqlite.Connection = Depends(get_db),
) -> PublicTaskResponse:
    service = TaskService(conn)
    try:
        task = await service.start_task(task_id)
    except ValueError as exc:
        if "not found" in str(exc).lower():
            raise HTTPException(status_code=404, detail="Task not found") from exc
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    await log_audit_event(conn, task.get("session_id"), "api", "task.started", task_id)
    await conn.commit()
    return _task_response(task)


@router.post("/{task_id}/cancel", response_model=PublicTaskResponse)
async def cancel_task(
    task_id: str,
    conn: aiosqlite.Connection = Depends(get_db),
) -> PublicTaskResponse:
    service = TaskService(conn)
    try:
        task = await service.cancel_task(task_id)
    except ValueError as exc:
        if "not found" in str(exc).lower():
            raise HTTPException(status_code=404, detail="Task not found") from exc
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    await log_audit_event(conn, task.get("session_id"), "api", "task.cancelled", task_id)
    await conn.commit()
    return _task_response(task)


@router.post("/{task_id}/actions", response_model=PublicTaskActionResponse, status_code=status.HTTP_201_CREATED)
async def request_task_action(
    task_id: str,
    request: PublicTaskActionCreateRequest,
    conn: aiosqlite.Connection = Depends(get_db),
) -> PublicTaskActionResponse:
    service = TaskService(conn)
    try:
        task = await service.request_approval(
            task_id,
            request.tool_name,
            request.description,
            risk_level=request.risk_level,
        )
    except ValueError as exc:
        if "not found" in str(exc).lower():
            raise HTTPException(status_code=404, detail="Task not found") from exc
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except TransitionError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc

    actions = await service.get_task_with_events(task_id)
    action = actions["actions"][-1]
    await log_audit_event(
        conn=conn,
        session_id=task.get("session_id"),
        actor="api",
        action="task_action.requested",
        target=action["id"],
        payload={"task_id": task_id, "risk_level": request.risk_level},
    )
    await conn.commit()
    return _action_response(action)


@router.post("/{task_id}/actions/{action_id}/decision", response_model=PublicTaskResponse)
async def decide_task_action(
    task_id: str,
    action_id: str,
    request: PublicTaskActionDecisionRequest,
    conn: aiosqlite.Connection = Depends(get_db),
) -> PublicTaskResponse:
    service = TaskService(conn)
    try:
        task = await service.resolve_approval(
            task_id,
            action_id,
            approved=request.approved,
            output_json=request.output_json,
        )
    except ValueError as exc:
        if "not found" in str(exc).lower():
            raise HTTPException(status_code=404, detail="Task not found") from exc
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except TransitionError as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc

    await log_audit_event(
        conn=conn,
        session_id=task.get("session_id"),
        actor="api",
        action="task_action.allowed" if request.approved else "task_action.denied",
        target=action_id,
        payload={"task_id": task_id},
    )
    await conn.commit()
    return _task_response(task)

File: frontend/src/api/tasks.ts
import { apiFetch } from './client';

export interface Task {
  id: string;
  session_id?: string | null;
  parent_task_id?: string | null;
  title?: string | null;
  description?: string | null;
  status: 'queued' | 'running' | 'waiting_approval' | 'succeeded' | 'failed' | 'cancelled';
  task_type: string;
  created_at: number;
  updated_at: number;
  duplicate: boolean;
}

export interface TaskEvent {
  id: string;
  task_id: string;
  run_id?: string | null;
  type: string;
  status: string;
  data_json?: string | null;
  created_at: number;
}

export interface TaskAction {
  id: string;
  task_id: string;
  tool_name: string;
  risk_level: string;
  status: string;
  description?: string | null;
  input_json?: string | null;
  output_json?: string | null;
  created_at: number;
  resolved_at?: number | null;
}

export interface TaskCreateRequest {
  session_id?: string | null;
  title?: string | null;
  description?: string | null;
  task_type?: string;
  parent_task_id?: string | null;
}

export interface TaskActionCreateRequest {
  tool_name: string;
  description: string;
  risk_level?: 'read' | 'write_internal' | 'external_or_destructive';
}

export interface TaskActionDecisionRequest {
  approved: boolean;
  output_json?: string | null;
}

export const createTask = async (
  request: TaskCreateRequest,
  idempotencyKey?: string,
): Promise<Task> => {
  const headers: Record<string, string> = {};
  if (idempotencyKey) {
    headers['Idempotency-Key'] = idempotencyKey;
  }
  return apiFetch<Task>('/api/tasks', {
    method: 'POST',
    headers,
    body: JSON.stringify(request),
  });
};

export const listTasks = async (params?: {
  session_id?: string;
  status?: string;
  limit?: number;
  offset?: number;
}): Promise<Task[]> => {
  const searchParams = new URLSearchParams();
  if (params?.session_id) searchParams.append('session_id', params.session_id);
  if (params?.status) searchParams.append('status', params.status);
  if (params?.limit !== undefined) searchParams.append('limit', params.limit.toString());
  if (params?.offset !== undefined) searchParams.append('offset', params.offset.toString());

  const query = searchParams.toString();
  return apiFetch<Task[]>(`/api/tasks${query ? `?${query}` : ''}`);
};

export const getTask = async (taskId: string): Promise<Task> => {
  return apiFetch<Task>(`/api/tasks/${taskId}`);
};

export const listTaskEvents = async (taskId: string): Promise<TaskEvent[]> => {
  return apiFetch<TaskEvent[]>(`/api/tasks/${taskId}/events`);
};

export const startTask = async (taskId: string): Promise<Task> => {
  return apiFetch<Task>(`/api/tasks/${taskId}/start`, {
    method: 'POST',
  });
};

export const cancelTask = async (taskId: string): Promise<Task> => {
  return apiFetch<Task>(`/api/tasks/${taskId}/cancel`, {
    method: 'POST',
  });
};

export const requestTaskAction = async (
  taskId: string,
  request: TaskActionCreateRequest,
): Promise<TaskAction> => {
  return apiFetch<TaskAction>(`/api/tasks/${taskId}/actions`, {
    method: 'POST',
    body: JSON.stringify(request),
  });
};

export const decideTaskAction = async (
  taskId: string,
  actionId: string,
  request: TaskActionDecisionRequest,
): Promise<Task> => {
  return apiFetch<Task>(`/api/tasks/${taskId}/actions/${actionId}/decision`, {
    method: 'POST',
    body: JSON.stringify(request),
  });
};
